wpiutil/src/main/native/cpp/DataLog.cpp

```diff
@@ -0,0 +1,771 @@
+    return rv;
+  }
+
+  void Unreserve(size_t size) { m_len -= size; }
+
+  void Clear() { m_len = 0; }
+
+  size_t GetRemaining() const { return m_maxLen - m_len; }
+
+  wpi::span<uint8_t> GetData() { return {m_buf, m_len}; }
+  wpi::span<const uint8_t> GetData() const { return {m_buf, m_len}; }
+
+ private:
+  uint8_t* m_buf;
+  size_t m_len = 0;
+  size_t m_maxLen;
+};
+
+DataLog::DataLog(std::string_view dir, std::string_view filename, double period,
+                 std::string_view extraHeader)
+    : m_period{period},
+      m_extraHeader{extraHeader},
+      m_newFilename{filename},
+      m_thread{[this, dir = std::string{dir}] { WriterThreadMain(dir); }} {}
+
+DataLog::DataLog(std::function<void(wpi::span<const uint8_t> data)> write,
+                 double period, std::string_view extraHeader)
+    : m_period{period},
+      m_extraHeader{extraHeader},
+      m_thread{[this, write = std::move(write)] {
+        WriterThreadMain(std::move(write));
+      }} {}
+
+DataLog::~DataLog() {
+  {
+    std::scoped_lock lock{m_mutex};
+    m_active = false;
+    m_doFlush = true;
+  }
+  m_cond.notify_all();
+  m_thread.join();
+}
+
+void DataLog::SetFilename(std::string_view filename) {
+  {
+    std::scoped_lock lock{m_mutex};
+    m_newFilename = filename;
+  }
+  m_cond.notify_all();
+}
+
+void DataLog::Flush() {
+  {
+    std::scoped_lock lock{m_mutex};
+    m_doFlush = true;
+  }
+  m_cond.notify_all();
+}
+
+void DataLog::Pause() {
+  std::scoped_lock lock{m_mutex};
+  m_paused = true;
+}
+
+void DataLog::Resume() {
+  std::scoped_lock lock{m_mutex};
+  m_paused = false;
+}
+
+static void WriteToFile(fs::file_t f, wpi::span<const uint8_t> data,
+                        std::string_view filename) {
+  do {
+#ifdef _WIN32
+    DWORD ret;
+    if (!WriteFile(f, data.data(), data.size(), &ret, nullptr)) {
+      fmt::print(stderr, "DataLog: Error writing to log file '{}': {}\n",
+                 filename, GetLastError());
+      break;
+    }
+#else
+    ssize_t ret = ::write(f, data.data(), data.size());
+    if (ret < 0) {
+      // If it's a recoverable error, swallow it and retry the write
+      if (errno == EINTR || errno == EAGAIN || errno == EWOULDBLOCK) {
+        continue;
+      }
+
+      // Otherwise it's a non-recoverable error; quit trying
+      fmt::print(stderr, "DataLog: Error writing to log file '{}': {}\n",
+                 filename, std::strerror(errno));
+      break;
+    }
+#endif
+
+    // The write may have written some or all of the data
+    data = data.subspan(ret);
+  } while (data.size() > 0);
+}
+
+void DataLog::WriterThreadMain(std::string_view dir) {
+  std::chrono::duration<double> periodTime{m_period};
+
+  std::error_code ec;
+  fs::path dirPath{dir};
+  std::string filename;
+
+  {
+    std::scoped_lock lock{m_mutex};
+    filename = std::move(m_newFilename);
+    m_newFilename.clear();
+  }
+
+  if (filename.empty()) {
+    // build random filename
+    static std::random_device dev;
```

Comment:
This should retry if there's an accidental collision?

Reply:
The probability of an accidental collision seems pretty low, since there are 16^16 combinations.  I guess the concern would be in an embedded environment running from startup that the random device wouldn't be well seeded yet?

Reply:
I feel like most temporary file implementations have a retry loop, and it's an easy enough thing to add.

Reply:
Re-reading the code, we simply replace any existing file.  So the only way this could fail would be a permissions issue, which is not likely to be fixed by a retry with a different filename?

Reply:
If the intent is to replace the old data, that seems fine. I would think that you aren't trying to discard old data though?

Reply:
We could make it so that for randomly generated filenames, it will retry a few times, but for a user-specified filename, it will overwrite?  There is actually different behavior here for the *initial* filename versus a *change* in filename after the thread is running.  The first will overwrite but the second will fail to rename, and thus keep the temp filename.  What is the "most correct" behavior?  If we want to never overwrite, it would mean that trying to write to an existing file will instead write to some other (random) filename.  That maybe is the best option?